kratos/utilities/math_utils.h

```diff
@@ -1574,45 +1574,120 @@ class MathUtils
      }
 
     /**
-     * @brief Calculates the eigenvectors and eigenvalues of given symmetric TDimxTDim matrix
-     * @details The eigenvectors and eigenvalues are calculated using the iterative Gauss-Seidel-method
+     * @brief Calculates the product operation B'DB
+     * @param rA The resulting matrix
+     * @param rD The "center" matrix
+     * @param rB The matrices to be transposed
+     * @tparam TMatrixType1 The type of matrix considered (1)
+     * @tparam TMatrixType2 The type of matrix considered (2)
+     * @tparam TMatrixType3 The type of matrix considered (3)
+     */
+    template<class TMatrixType1, class TMatrixType2, class TMatrixType3>
+    static inline void BtDBProductOperation(
+        TMatrixType1& rA,
+        const TMatrixType2& rD,
+        const TMatrixType3& rB
+        )
+    {
+        // The sizes
+        const SizeType size1 = rB.size2();
+        const SizeType size2 = rB.size2();
+
+#ifdef KRATOS_USE_AMATRIX   // This macro definition is for the migration period and to be removed afterward please do not use it
+        KRATOS_WARNING_IF("BtDBProductOperation", rA.size1() != size1 || rA.size2() != size2) << "BtDBProductOperation has detected an incorrect size of your resulting matrix matrix. Please resize before compute" << std::endl;
+#else
+        if (rA.size1() != size1 || rA.size2() != size2)
+            rA.resize(size1, size2, false);
+#endif // KRATOS_USE_AMATRIX
+
+        noalias(rA) = prod( trans( rB ), MatrixType(prod(rD, rB)));
```

Comment:
Controlling the complexity of nested products

What is the complexity (the number of add and multiply operations) required to compute the following?

 R = prod(A, prod(B,C)); 

Firstly the complexity depends on matrix size. Also since prod is transitive (not commutative) the bracket order affects the complexity.

uBLAS evaluates expressions without matrix or vector temporaries and honours the bracketing structure. However avoiding temporaries for nested product unnecessarly increases the complexity. Conversly by explictly using temporary matrices the complexity of a nested product can be reduced.

uBLAS provides 3 alternative syntaxes for this purpose:

 temp_type T = prod(B,C); R = prod(A,T);   // Preferable if T is preallocated

 prod(A, temp_type(prod(B,C));

 prod(A, prod<temp_type>(B,C));

@RiccardoRossi

Reply:
@loumalouomega to my understanding

prod<temp_type>(B,C)

and

prod(B,C)

are completely equivalent, and both create an expression. am i wrong?

in any case your solution with the function BtDB and BDBt is way superior in terms of performance (if you care of writing correctly the loops for cache efficiency)

Reply:
I already did. You can check it out